File: python/pokesim/move.py
"""
Defines data structures and routines for dealing with a pokemon's move
"""


import random
import typing
import enum
import os
from . import constants
from . import poketypes
from . import utils


def critical(movecrit: int, pokestages:int) -> float:
	"""
	Given a move's inherent critical hit ratio and a pokemon's critical stage,
	returns a multiplier to apply to damage from critical hits
	"""
	effectivestage = (movecrit + pokestages)

	if effectivestage > 3:
		return 1.5

	ratio = 1.0 / (2 ** (4-effectivestage))
	critChance = random.uniform(0,1)

	return 1.5 if ratio > critChance else 1

class MoveType(enum.IntEnum):
	"""
	The type of a move
	"""
	Status   = 0
	Special  = 1
	Physical = 2

	def __str__(self) -> str:
		"""
		Returns the name of a move type
		"""
		return self._name_

# Move Type constants for export
STATUS = MoveType.Status
SPECIAL = MoveType.Special
PHYSICAL = MoveType.Physical

class Move():
	"""
	A pokemon's move
	"""
	def __init__(self, name: str):
		"""
		Reads in a move's data. Expects it in `../data/moves/<name>`
		"""
		self.name = name
		self.priority = 0

		with open(os.path.join(utils.dataDir, "moves", name)) as datafile:
			lines = datafile.read().split('\n')
			_ = lines.pop() #eliminates POSIX-compliant empty line at file end
			self.type1 = int(lines.pop(0))
			self.type2 = int(lines.pop(0))
			self.contact = bool(int(lines.pop(0)))
			self.accuracy = int(lines.pop(0))
			self.PP = int(lines.pop(0))
			self.moveType = MoveType(int(lines.pop(0)))

		print(lines)

		if self.moveType != STATUS:
			self.damagingMove(lines)
		else:
			self.statusMove(lines)

	def damagingMove(self, lines: typing.List[str]):
		"""
		Builds a damaging move from the remaining lines in a file after common
		attributes have already been parsed out
		"""
		self.crit = int(lines.pop())
		self.power = int(lines.pop())

	def statusMove(self, lines: typing.List[str]):
		"""
		Builds a status move from the remaining lines in a file after common
		attributes have already been parsed out
		"""
		self.target = bool(int(lines.pop()))
		self.stageChange = int(lines.pop())
		self.affectedStat = int(lines.pop())

	def calcDmg(self, pkmn: object, otherpkmn: object, othermove: 'Move') -> int:
		"""
		Calculates damage done by pkmn to otherpkmn (who used 'othermove', if that matters)
		"""
		dmg = 2 * pkmn.level / 5.0
		dmg += 2
		dmg *= self.power

		crit = critical(self.crit, pkmn.stages[constants.CRIT])
		if crit > 1:
			print("Critical Hit!")

		effat, effdef = 0.0, 0.0
		atstage, defstage = None, None

		#Physical attack
		if self.moveType == PHYSICAL:
			effat = pkmn.attack
			effdef = pkmn.defense
			atstage = pkmn.stages[constants.ATTACK]
			defstage = otherpkmn.stages[constants.DEFENSE]
		#Special attack
		else:
			effat = pkmn.specialAttack
			effdef = pkmn.specialDefense
			atstage = pkmn.stages[constants.SPECIAL_ATTACK]
			defstage = otherpkmn.stages[constants.SPECIAL_DEFENSE]

		if atstage > 0:
			effat *= (2.0 + atstage) / 2.0
		elif atstage < 0 and crit == 1:
			effat *= 2.0 / (2.0 - atstage)

		if defstage > 0 and crit ==1:
			effdef *= (2.0 + defstage) / 2.0
		elif defstage < 0:
			effdef *= 2.0 / (2.0 - defstage)


		dmg *= effat/effdef
		dmg /= 50
		dmg += 2

		#Caclucate modifier
		mod = random.uniform(0.85, 1)
		if self.type1 != poketypes.TYPELESS:
			if pkmn.type1 == self.type1 or pkmn.type2 == self.type2:
				mod *= 1.5

		if self.type2 != poketypes.TYPELESS:
			if pkmn.type1 == self.type2 or pkmn.type2 == self.type2:
				mod *= 1.5

		if pkmn.status == constants.BRN and self.moveType == PHYSICAL:
			mod /= 2.0

		dmg *= mod * crit

		return int(dmg * mod)

	def __repr__(self) -> str:
		"""
		Forms an in-depth representation of a Move
		"""
		printstr = "%s\nType: %s%s\n" % (self, self.type1,
		                       ('/'+str(self.type2) if self.type2 != poketypes.TYPELESS else ''))
		printstr += "%s\nAccuracy: %s\n" % (self.moveType, self.accuracy)
		if self.moveType != STATUS:
			printstr += "Power: %s\nDoes %smake contact\n" % (self.power,
			                                                 ('not' if not self.contact else ''))
		return printstr

	def __str__(self) -> str:
		"""
		Returns the human-readable name of a Move
		"""
		return self.name

File: javascript.node/move.ts
import { join as pathJoin } from "path";
import { promises as fs } from "fs";

import { DATA_DIR, type StageableStat, parseStatList } from "./constants";
import { Type, parseType, printType } from "./poketypes";
import { parseBool, parseInteger, parseIntegerList } from "./ioutils";

/**
 * Gets a multiplier for critical hitting.
 *
 * @param movecrit The used move's inherent critical hit ratio.
 * @param pokestage The critical hit ratio "stage" of the Pokémon using the
 * move.
 * @returns A multiplier to apply to damage based on whether or not the move
 * crits.
 */
export function critical(movecrit: number, pokestages: number): number {
	const effectivestage = movecrit + pokestages;

	if (effectivestage > 3) {
		return 1.5;
	}

	const ratio = 1.0 / (2 ** (4-effectivestage))

	if (ratio > Math.random()) {
		return 1.5
	}
	return 1
}

/** The "kind" of a Pokémon's Move. */
export const enum MoveKind {
	/** A move that doesn't deal direct damage. */
	Status   = 0,
	/**
	 * A directly damaging move that doesn't make contact. Uses Special Attack
	 * and Special Defense for damage calculation.
	 */
	Special  = 1,
	/**
	 * A directly damaging move that makes contact. Uses Attack and Defense for
	 * damage calculation.
	 */
	Physical = 2
};

/** A mapping of the "kinds" of Pokémon Moves to their names. */
export const moveKindName: Readonly<Record<MoveKind, string>> = {
	[MoveKind.Status]: "Status",
	[MoveKind.Special]: "Special",
	[MoveKind.Physical]: "Physical"
};

/**
 * Attempts to construct a MoveKind from a raw number.
 *
 * @param input The number to be parsed. This should be equivalent to the enum
 * member of the MoveKind it represents.
 * @returns The MoveKind represented.
 * @throws {Error} if the given raw MoveKind cannot be parsed, and/or parses to
 * an integer that doesn't represent a MoveKind.
 */
export function parseMoveKind(input: number): MoveKind {
	switch(input) {
		case MoveKind.Physical:
		case MoveKind.Special:
		case MoveKind.Status:
			return input;
	}

	throw new Error(`unrecognized internal MoveKind representation: ${input}`);
}

/**
 * A move that is or can be known by a Pokémon.
 */
export class Move {
	public priority = 0;

	public type1 = Type.Typeless;
	public type2 = Type.Typeless;
	public accuracy = 100;
	public PP = 0;
	public maxPP = 0;
	public kind = MoveKind.Status;
	/** @todo Redundant? */
	public contact = false;

	/**
	 * Whether or not the move is considered a "shadow" move for the purposes of
	 * type effectiveness.
	 */
	public get shadow(): boolean {
		return this.type1 === Type.Shadow || this.type2 === Type.Shadow;
	}

	// properties that only have meaning for damaging moves.
	public crit = 0;
	public power = 0;

	// properties that only have meaning for status moves (for now).
	public target = false;
	public stageChanges = new Array<[StageableStat, number]>();

	constructor(public readonly name: string) { }

	/**
	 * Reads in a move's data from the data directory under `moves/{{name}}`.
	 */
	public async readData(): Promise<void> {
		const filename = pathJoin(DATA_DIR, "moves", this.name);
		const lines = (await fs.readFile(filename, {encoding: "utf-8"})).split("\n");

		let i = 0;
		function nextLine(): string | undefined {
			++i;
			return lines.shift();
		}
		try {
			this.type1 = parseType(parseInteger(nextLine()));
			this.type2 = parseType(parseInteger(nextLine()));
			this.contact = parseBool(nextLine());
			this.accuracy = parseInteger(nextLine(), true);
			this.PP = parseInteger(nextLine(), true);
			this.maxPP = this.PP;
			this.kind = parseMoveKind(parseInteger(nextLine()));

			if (this.kind !== MoveKind.Status) {
				this.power = parseInteger(nextLine(), true);
				this.crit = parseInteger(nextLine(), true);
			} else {
				const affectedStats = parseStatList(nextLine());
				const asl = affectedStats.length;

				const statChanges = parseIntegerList(nextLine());
				const scl = statChanges.length;
				if (asl !== scl) {
					throw new Error(`list of affected stats length (${asl}) does not match stat changes list length (${scl})`);
				}

				this.stageChanges = [];
				for (let sci = 0; sci < scl; ++sci) {
					this.stageChanges.push([affectedStats[sci], statChanges[sci]]);
				}

				this.target = parseBool(nextLine());
			}
		} catch (e) {
			throw new Error(`line #${i} of ${filename}: ${e instanceof Error ? e.message : e}`);
		}

	}

	/**
	 * Constructs a string representation of a Move meant to be viewed by
	 * developers.
	 */
	public debugStr(): string {
		let ret = [this, `Type: ${printType(this.type1)}`]
		if (this.type2 !== Type.Typeless) {
			ret[1] += `/${printType(this.type2)}`;
		}
		ret = ret.concat([moveKindName[this.kind], `Accuracy: ${this.accuracy}`]);
		if (this.kind !== MoveKind.Status) {
			ret = ret.concat([`Power: ${this.power}`, `Does ${this.contact ? "not make" : "make"} contact`]);
		}
		ret.push(`PP: ${this.PP}/${this.maxPP}`);
		return ret.join("\n");
	}

	/**
	 * Returns the human-readable name of a Move.
	 */
	public toString(): string {
		return this.name
	}
}
